crates/chia-datalayer/src/lib.rs

```diff
@@ -0,0 +1,1259 @@
+        old_leaf: &Node,
+        internal_node_hash: Hash,
+    ) -> Result<(), String> {
+        let new_leaf_index = self.get_new_index();
+        let new_internal_node_index = self.get_new_index();
+
+        let new_leaf_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Leaf,
+                dirty: false,
+            },
+            node: Node {
+                parent: Some(new_internal_node_index),
+                specific: NodeSpecific::Leaf { key_value },
+                hash,
+                index: new_leaf_index,
+            },
+        };
+        self.insert_entry_to_blob(new_leaf_index, new_leaf_block.to_bytes())?;
+
+        let new_internal_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Internal,
+                dirty: false,
+            },
+            node: Node {
+                parent: old_leaf.parent,
+                specific: NodeSpecific::Internal {
+                    left: old_leaf.index,
+                    right: new_leaf_index,
+                },
+                hash: internal_node_hash,
+                index: new_internal_node_index,
+            },
+        };
+        self.insert_entry_to_blob(new_internal_node_index, new_internal_block.to_bytes())?;
+
+        let Some(old_parent_index) = old_leaf.parent else {
+            panic!("{key_value:?} {hash:?}")
+        };
+
+        let mut block = Block::from_bytes(
+            self.get_block_bytes(old_leaf.index)?,
+            new_internal_node_index,
+        )?;
+        block.node.parent = Some(new_internal_node_index);
+        self.insert_entry_to_blob(old_leaf.index, block.to_bytes())?;
+
+        let mut old_parent_block =
+            Block::from_bytes(self.get_block_bytes(old_parent_index)?, old_parent_index)?;
+        match old_parent_block.node.specific {
+            NodeSpecific::Internal {
+                ref mut left,
+                ref mut right,
+                ..
+            } => {
+                if old_leaf.index == *left {
+                    *left = new_internal_node_index;
+                } else if old_leaf.index == *right {
+                    *right = new_internal_node_index;
+                } else {
+                    panic!();
+                }
+            }
+            NodeSpecific::Leaf { .. } => panic!(),
+        }
+        self.insert_entry_to_blob(old_parent_index, old_parent_block.to_bytes())?;
+
+        self.mark_lineage_as_dirty(old_parent_index)?;
+        self.kv_to_index.insert(key_value, new_leaf_index);
+
+        Ok(())
+    }
+
+    pub fn delete(&mut self, key_value: KvId) -> Result<(), String> {
+        let leaf_index = *self.kv_to_index.get(&key_value).unwrap();
+        let leaf = self.get_node(leaf_index).unwrap();
+
+        match leaf.specific {
+            // TODO: blech
+            NodeSpecific::Leaf { .. } => (),
+            NodeSpecific::Internal { .. } => panic!(),
+        };
```

Comment:
I think this could be something like:
```
assert!(matches!(NodeSpecific::Leaf { ... }, leaf.specific));
```
But if there's a way to make the code just type-safe by construction, that would be ideal I think.

Reply:
i bet you can imagine the python source that i was copying when i wrote this.  `:]`  type narrowing and all.  but yes, the `blech` is saying i don't like this and should revisit to try to do something else.  i was definitely hitting this due to some bugs though.

Reply:
I think this is all gone at this point.  Or at least wrapped up a bit.  https://github.com/Chia-Network/chia_rs/pull/673/files#diff-a48504f4b97458ce535cd911838701acf1c6b739aeb9feac35a52688c432ec6aR776-R828